go/pkg/client/cas.go

```diff
@@ -25,13 +27,227 @@ import (
+	actionID     string
+	invocationID string
+}
+
+type uploadRequest struct {
+	c    *chunker.Chunker
+	meta *requestMetadata
+	wait chan<- *uploadResponse
+}
+
+type uploadResponse struct {
+	digest  digest.Digest
+	err     error
+	missing bool
+}
+
+type uploadState struct {
+	data *chunker.Chunker
+	err  error
+
+	// mu protects clients. It does NOT protect data or error, because they do not need protection -
+	// they are only modified when a state object is created, and by updateAndNotify which is called
+	// exactly once for a given state object (this is the whole point of the algorithm).
+	mu      sync.Mutex
+	clients []chan<- *uploadResponse
+}
+
+func (c *Client) findBlobState(ctx context.Context, dgs []digest.Digest) (missing []digest.Digest, present []digest.Digest, err error) {
+	dgMap := make(map[digest.Digest]bool)
+	for _, d := range dgs {
+		dgMap[d] = true
+	}
+	missing, err = c.MissingBlobs(ctx, dgs)
+	for _, d := range missing {
+		delete(dgMap, d)
+	}
+	for d := range dgMap {
+		present = append(present, d)
+	}
+	return missing, present, err
+}
+
+func (c *Client) uploadProcessor() {
+	var buffer []*uploadRequest
+	ticker := time.NewTicker(50 * time.Millisecond)
+	for {
+		select {
+		case req, ok := <-c.casUploadRequests:
+			if !ok {
+				// Client is exiting. Notify remaining uploads to prevent deadlocks.
+				ticker.Stop()
+				if buffer != nil {
+					for _, r := range buffer {
+						r.wait <- &uploadResponse{err: context.Canceled}
+					}
+				}
+				return
+			}
+			buffer = append(buffer, req)
+			if len(buffer) >= casChanBufferSize {
+				c.upload(buffer)
+				buffer = nil
+			}
+		case <-ticker.C:
+			if buffer != nil {
+				c.upload(buffer)
+				buffer = nil
+			}
+		}
+	}
+}
+
+func updateAndNotify(st *uploadState, err error, missing bool) {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+	st.err = err
+	for i, cl := range st.clients {
+		cl <- &uploadResponse{
+			digest:  st.data.Digest(),
+			missing: i == 0 && missing, // Only first client is reported the digest missing.
+			err:     err,
+		}
+	}
+	st.clients = nil
+	st.data = nil
+}
+
+func getUnifiedMetadata(metas []*requestMetadata) *requestMetadata {
+	toolNames := make(map[string]bool)
+	actionIDs := make(map[string]bool)
+	invocationIDs := make(map[string]bool)
+	for _, m := range metas {
+		toolNames[m.toolName] = true
+		actionIDs[m.actionID] = true
+		invocationIDs[m.invocationID] = true
+	}
+	var tnKeys, aiKeys, iiKeys []string
+	for toolName := range toolNames {
+		tnKeys = append(tnKeys, toolName)
+	}
+	for actionID := range actionIDs {
+		aiKeys = append(aiKeys, actionID)
+	}
+	for invocationID := range invocationIDs {
+		iiKeys = append(iiKeys, invocationID)
+	}
+	sort.Strings(tnKeys)
+	sort.Strings(aiKeys)
+	sort.Strings(iiKeys)
+	return &requestMetadata{
+		toolName:     strings.Join(tnKeys, ","),
+		actionID:     strings.Join(aiKeys, ","),
+		invocationID: strings.Join(iiKeys, ","),
+	}
+}
+
+func (c *Client) upload(reqs []*uploadRequest) {
+	const (
```

Comment:
nit: declare it at the top

Reply:
Done. I should just implement log.EveryN()...